aten/src/ATen/SparseTensorImpl.h

```diff
@@ -210,6 +210,32 @@ struct CAFFE2_API SparseTensorImpl : public TensorImpl {
     impl->refresh_numel();
     return impl;
   }
+
+  // NOTE: `shallow_copy_from()` does not copy the AutogradMeta pointer
+  // because it is unique for each Variable.
+  virtual void shallow_copy_from(c10::intrusive_ptr<TensorImpl> impl) override {
+    AT_ASSERT(impl->is_sparse());
+    auto sparse_impl = static_cast<SparseTensorImpl*>(impl.get());
+    type_id_ = sparse_impl->type_id();
+    data_type_ = sparse_impl->dtype();
+
+    // TensorImpl general fields
+    // Note that these fields are not used in sparse tensor code, and we copy them here only for completeness.
+    sizes_ = sparse_impl->sizes_;
+    strides_ = sparse_impl->strides_;
+    storage_offset_ = sparse_impl->storage_offset_;
+    is_contiguous_ = sparse_impl->is_contiguous_;
+    is_wrapped_number_ = sparse_impl->is_wrapped_number_;
+    reserved_ = sparse_impl->reserved_;
```

Comment:
can we have a function that does the dense part (that's smart wrt e.g. not calling contiguous) in one place?